shared/src/main/scala/eu/timepit/refined/api/Validate.scala

```diff
@@ -0,0 +1,113 @@
+package eu.timepit.refined
+package api
+
+import eu.timepit.refined.internal.Resources
+
+import scala.util.Try
+
+/**
+ * Type class for validating values of type `T` according to a type-level
+ * predicate `P`. The semantics of `P` are defined by the instance(s) of
+ * this type class for `P`.
+ */
+trait Validate[T, P] extends Serializable { self =>
+
+  type R
+
+  final type Res = Result[R]
+
+  def validate(t: T): Res
+
+  /** Returns a string representation of this [[Validate]] using `t`. */
+  def showExpr(t: T): String
```

Comment:
I would prefer to have a type class (`Show` from `cats` for instance) instead of `showExpr` and `showResult`